stablehlo/dialect/StablehloOps.cpp

```diff
@@ -2294,148 +2294,112 @@ LogicalResult AllGatherOp::verify() {
-                         << operandType.getRank() << " and rank(grad_output) "
-                         << gradOutputType.getRank() << ".";
-
-  // C3.
-  const int64_t featureCount = operandType.getShape()[feature_index()];
+LogicalResult verifyBatchNorm(Location loc, Value operand,
+                              int64_t feature_index, Value scale) {
+  auto operandType = operand.getType().cast<RankedTensorType>();
+  if (static_cast<int64_t>(feature_index) >= operandType.getRank())
+    return emitError(loc) << "expects feature_index to be smaller "
+                             "than the rank of operand type; got feature_index "
+                          << feature_index << ", and rank "
+                          << operandType.getRank() << ".";
+
+  if (static_cast<int64_t>(feature_index) < 0)
+    return emitError(loc) << "expects feature_index to be a "
+                          << "non-negative number, got "
+                          << static_cast<int64_t>(feature_index) << ".";
+
+  const int64_t featureCount = operandType.getShape()[feature_index];
   const int64_t scaleShape =
-      scale().getType().cast<RankedTensorType>().getShape()[0];
+      scale.getType().cast<RankedTensorType>().getShape()[0];
+  // As ODS enforces `scale`, `mean`, `variance`, `offset` are AllShapesMatch,
+  // this also infers that featureCount is aligned with them.
   if (scaleShape != featureCount)
-    return emitOpError() << "expects the size of scale factor to be "
-                            "same as the feature count,"
-                            " but the size of scale factor is "
-                         << scaleShape << " and the feature count is "
-                         << featureCount << ".";
+    return emitError(loc) << "expects the size of scale factor to be "
+                             "same as the feature count,"
+                             " but the size of scale factor is "
+                          << scaleShape << " and the feature count is "
+                          << featureCount << ".";
+
+  return success();
+}
+
+// Refer ODS for properties that are already enforced including shapes and
+// element types. This verifier includes additional checks.
+LogicalResult BatchNormGradOp::verify() {
+  if (failed(verifyBatchNorm(getLoc(), operand(), feature_index(), scale())))
+    return failure();
+  return success();
+}
+
+LogicalResult BatchNormGradOp::inferReturnTypeComponents(
+    MLIRContext* context, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  BatchNormGradOp::Adaptor adaptor(operands, attributes, regions);
+
+  RankedTensorType operandType =
+      adaptor.operand().getType().cast<RankedTensorType>();
+  inferredReturnShapes.emplace_back(operandType.getShape(),
+                                    operandType.getElementType());
```

Comment:
I think this can be simplified to `inferredReturnShapes.emplace_back(operandType)` because `operandType` is a `ShapedType`. (Here and below).

Reply:
This looks better but actually not working. Yes `operandType` is a `ShapedType`, but it is also a `Type`. `ShapedTypeComponents` (the element type of `inferredReturnShapes` has contructors:
```
ShapedTypeComponents(Type elementType) ...
ShapedTypeComponents(ShapedType shapedType) : attr(nullptr) ...
```
So the Type one is used rather than what we expect. As I see, `ShapedType` is not a `Type`. But our `operandType` is fit for both constructors, so it ends up to be used as elementType and error out.

Reply:
What I meant is that we can `.cast<ShapedType>()` and then the simpler `emplace_back` will work.

Reply:
Got it. That's clean and fixed.